{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b74ea615-3cef-4e47-8376-22b75a206dde",
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import pandas as pd\n",
    "\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac6e52b4-866a-4e1a-89ed-d34c97fbc3df",
   "metadata": {},
   "source": [
    "# Load in SRA metadata and search results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80e3f77e-3989-4973-b97f-7bb5d9d27e9f",
   "metadata": {},
   "source": [
    "### Load in SRA metadata info:\n",
    "\n",
    "This is the 'run info' that you can download from NCBI in bulk; it's got one entry for every accession, approximately."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9ce1dbc3-35b2-428b-839d-7e1c0381f587",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/1c/s23lyl9x7cgd1rg4n6t867qr0000gn/T/ipykernel_10142/1351821980.py:1: DtypeWarning: Columns (8,33,38,39,40,43) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  run_info = pd.read_csv('../big.runinfo.csv.gz')\n"
     ]
    }
   ],
   "source": [
    "run_info = pd.read_csv('../big.runinfo.csv.gz')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6edeb8a6-e8ec-4ce8-b7b2-541e4c700448",
   "metadata": {},
   "source": [
    "The two most important columns for our purposes are 'Run' and 'ScientificName':"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "58b6516d-5261-4758-a7d4-b78e61b6b6d9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['Run', 'ReleaseDate', 'LoadDate', 'spots', 'bases', 'spots_with_mates',\n",
      "       'avgLength', 'size_MB', 'AssemblyName', 'download_path', 'Experiment',\n",
      "       'LibraryName', 'LibraryStrategy', 'LibrarySelection', 'LibrarySource',\n",
      "       'LibraryLayout', 'InsertSize', 'InsertDev', 'Platform', 'Model',\n",
      "       'SRAStudy', 'BioProject', 'Study_Pubmed_id', 'ProjectID', 'Sample',\n",
      "       'BioSample', 'SampleType', 'TaxID', 'ScientificName', 'SampleName',\n",
      "       'g1k_pop_code', 'source', 'g1k_analysis_group', 'Subject_ID', 'Sex',\n",
      "       'Disease', 'Tumor', 'Affection_Status', 'Analyte_Type',\n",
      "       'Histological_Type', 'Body_Site', 'CenterName', 'Submission',\n",
      "       'dbgap_study_accession', 'Consent', 'RunHash', 'ReadHash'],\n",
      "      dtype='object')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>ScientificName</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR18036904</td>\n",
       "      <td>bovine metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR18036905</td>\n",
       "      <td>bovine metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR18036906</td>\n",
       "      <td>bovine metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR18036907</td>\n",
       "      <td>bovine metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR18036908</td>\n",
       "      <td>bovine metagenome</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Run     ScientificName\n",
       "0  SRR18036904  bovine metagenome\n",
       "1  SRR18036905  bovine metagenome\n",
       "2  SRR18036906  bovine metagenome\n",
       "3  SRR18036907  bovine metagenome\n",
       "4  SRR18036908  bovine metagenome"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(run_info.columns)\n",
    "\n",
    "run_info[['Run', 'ScientificName']].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "588bc0f9-9293-43c6-b2c9-ec999241789f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "702013"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# there are ~700,000 entries:\n",
    "len(run_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "003061ca-fca4-46c0-90c0-3b7a63989e85",
   "metadata": {},
   "outputs": [],
   "source": [
    "run_info2 = run_info[['Run', 'ScientificName']]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e32d908d-dcf9-40e0-ac34-03683ec98cb0",
   "metadata": {},
   "source": [
    "### Now, load in the stamps MAGsearch results\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e1f1e94a-71f7-4d23-9894-0a69dfa89237",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5527\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query</th>\n",
       "      <th>Run</th>\n",
       "      <th>containment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FV_DSM_15829_genome</td>\n",
       "      <td>/group/ctbrowngrp/irber/data/wort-data/wort-sr...</td>\n",
       "      <td>0.038168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FV_DSM_15829_genome</td>\n",
       "      <td>/group/ctbrowngrp/irber/data/wort-data/wort-sr...</td>\n",
       "      <td>0.091603</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FV_PB189-T1-4_genome</td>\n",
       "      <td>/group/ctbrowngrp/irber/data/wort-data/wort-sr...</td>\n",
       "      <td>0.037313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FV_DSM_15829_genome</td>\n",
       "      <td>/group/ctbrowngrp/irber/data/wort-data/wort-sr...</td>\n",
       "      <td>0.351145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FV_PB189-T1-4_genome</td>\n",
       "      <td>/group/ctbrowngrp/irber/data/wort-data/wort-sr...</td>\n",
       "      <td>0.149254</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  query                                                Run  \\\n",
       "0   FV_DSM_15829_genome  /group/ctbrowngrp/irber/data/wort-data/wort-sr...   \n",
       "1   FV_DSM_15829_genome  /group/ctbrowngrp/irber/data/wort-data/wort-sr...   \n",
       "2  FV_PB189-T1-4_genome  /group/ctbrowngrp/irber/data/wort-data/wort-sr...   \n",
       "3   FV_DSM_15829_genome  /group/ctbrowngrp/irber/data/wort-data/wort-sr...   \n",
       "4  FV_PB189-T1-4_genome  /group/ctbrowngrp/irber/data/wort-data/wort-sr...   \n",
       "\n",
       "   containment  \n",
       "0     0.038168  \n",
       "1     0.091603  \n",
       "2     0.037313  \n",
       "3     0.351145  \n",
       "4     0.149254  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "magsearch_df = pd.read_csv('../output.magsearch.k31/results/stamps.csv', quotechar=\"'\")\n",
    "print(len(magsearch_df))\n",
    "magsearch_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca990e2d-f687-4ba6-a967-6237e82c911c",
   "metadata": {},
   "source": [
    "# Make the results more human-readable and add SRA metadata info"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c32dd396-308f-49e0-b37a-8a1ef5e278a6",
   "metadata": {},
   "source": [
    "### First, we need to take the filenames in the 'Run' column and turn them into accessions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "44355ee8-d8fb-492a-ad1a-cdac5a4728ec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query</th>\n",
       "      <th>Run</th>\n",
       "      <th>containment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FV_DSM_15829_genome</td>\n",
       "      <td>ERR4333983</td>\n",
       "      <td>0.038168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FV_DSM_15829_genome</td>\n",
       "      <td>SRR17547361</td>\n",
       "      <td>0.091603</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FV_PB189-T1-4_genome</td>\n",
       "      <td>SRR17547361</td>\n",
       "      <td>0.037313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FV_DSM_15829_genome</td>\n",
       "      <td>ERR2014370</td>\n",
       "      <td>0.351145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FV_PB189-T1-4_genome</td>\n",
       "      <td>ERR2014370</td>\n",
       "      <td>0.149254</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  query          Run  containment\n",
       "0   FV_DSM_15829_genome   ERR4333983     0.038168\n",
       "1   FV_DSM_15829_genome  SRR17547361     0.091603\n",
       "2  FV_PB189-T1-4_genome  SRR17547361     0.037313\n",
       "3   FV_DSM_15829_genome   ERR2014370     0.351145\n",
       "4  FV_PB189-T1-4_genome   ERR2014370     0.149254"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def extract_run_acc(x):\n",
    "    # get just the end filename\n",
    "    x = os.path.basename(x)\n",
    "    # remove extension '.sig'\n",
    "    y, ext = os.path.splitext(x)\n",
    "    assert ext == '.sig', ext\n",
    "    return y\n",
    "\n",
    "# this can be used in case we have .gz, .fasta, .fa, etc in the query filename\n",
    "def remove_extension(x):\n",
    "    x = os.path.basename(x)\n",
    "    y, ext = os.path.splitext(x)\n",
    "    while ext in ('.gz', '.fasta', '.fa', '.fna'):\n",
    "        x = y\n",
    "        y, ext = os.path.splitext(x)\n",
    "    return y\n",
    "\n",
    "magsearch_df['Run'] = magsearch_df['Run'].apply(extract_run_acc)\n",
    "magsearch_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11456063-aeae-4a37-bf86-830924904b29",
   "metadata": {},
   "source": [
    "### Now we can correlate magsearch results with SRA RunInfo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "dfc9eb63-eff9-4c18-8738-d17d3718a80b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ScientificName</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Run</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR18036904</th>\n",
       "      <td>bovine metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR18036905</th>\n",
       "      <td>bovine metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR18036906</th>\n",
       "      <td>bovine metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR18036907</th>\n",
       "      <td>bovine metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR18036908</th>\n",
       "      <td>bovine metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR11108097</th>\n",
       "      <td>gut metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR8144073</th>\n",
       "      <td>Uvigerina striata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR8144074</th>\n",
       "      <td>Uvigerina striata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR8144079</th>\n",
       "      <td>Valvulineria inflata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR8144080</th>\n",
       "      <td>Valvulineria inflata</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>702013 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   ScientificName\n",
       "Run                              \n",
       "SRR18036904     bovine metagenome\n",
       "SRR18036905     bovine metagenome\n",
       "SRR18036906     bovine metagenome\n",
       "SRR18036907     bovine metagenome\n",
       "SRR18036908     bovine metagenome\n",
       "...                           ...\n",
       "SRR11108097        gut metagenome\n",
       "SRR8144073      Uvigerina striata\n",
       "SRR8144074      Uvigerina striata\n",
       "SRR8144079   Valvulineria inflata\n",
       "SRR8144080   Valvulineria inflata\n",
       "\n",
       "[702013 rows x 1 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "run_info2.set_index('Run')#['ScientificName']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "19db3333-a494-4d9c-b2c6-349c4e136e3a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query</th>\n",
       "      <th>containment</th>\n",
       "      <th>ScientificName</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Run</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DRR042391</th>\n",
       "      <td>FV_DSM_15829_genome</td>\n",
       "      <td>0.030534</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRR042545</th>\n",
       "      <td>FV_DSM_15829_genome</td>\n",
       "      <td>0.022901</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRR042550</th>\n",
       "      <td>FV_DSM_15829_genome</td>\n",
       "      <td>0.022901</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRR042658</th>\n",
       "      <td>FV_DSM_15829_genome</td>\n",
       "      <td>0.015267</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRR086622</th>\n",
       "      <td>S26</td>\n",
       "      <td>0.020896</td>\n",
       "      <td>marine metagenome</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         query  containment     ScientificName\n",
       "Run                                                           \n",
       "DRR042391  FV_DSM_15829_genome     0.030534                NaN\n",
       "DRR042545  FV_DSM_15829_genome     0.022901                NaN\n",
       "DRR042550  FV_DSM_15829_genome     0.022901                NaN\n",
       "DRR042658  FV_DSM_15829_genome     0.015267                NaN\n",
       "DRR086622                  S26     0.020896  marine metagenome"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "magsearch2_df = magsearch_df.set_index('Run').join(run_info2.set_index('Run')['ScientificName'])\n",
    "magsearch2_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "378dde5b-7208-4461-91ca-e7274df52c3c",
   "metadata": {},
   "source": [
    "### Subset to just SRA results with good scientific names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "fb2631c7-3dd5-4cf8-b8d7-0fb2230d9880",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "601\n"
     ]
    }
   ],
   "source": [
    "# how many have 'null' scientific name?\n",
    "null_df = magsearch2_df[magsearch2_df['ScientificName'].isnull()]\n",
    "print(len(null_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "11e7b3ae-d8db-46af-92b1-a37ebc962ead",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Of 5527 MAGsearch results, 4926 have non-null metadata (89.13%)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query</th>\n",
       "      <th>containment</th>\n",
       "      <th>ScientificName</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Run</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DRR086622</th>\n",
       "      <td>S26</td>\n",
       "      <td>0.020896</td>\n",
       "      <td>marine metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRR086623</th>\n",
       "      <td>S26</td>\n",
       "      <td>0.018140</td>\n",
       "      <td>marine metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRR086624</th>\n",
       "      <td>S26</td>\n",
       "      <td>0.019977</td>\n",
       "      <td>marine metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRR086640</th>\n",
       "      <td>S26</td>\n",
       "      <td>0.017451</td>\n",
       "      <td>marine metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRR086642</th>\n",
       "      <td>S26</td>\n",
       "      <td>0.024569</td>\n",
       "      <td>marine metagenome</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          query  containment     ScientificName\n",
       "Run                                            \n",
       "DRR086622   S26     0.020896  marine metagenome\n",
       "DRR086623   S26     0.018140  marine metagenome\n",
       "DRR086624   S26     0.019977  marine metagenome\n",
       "DRR086640   S26     0.017451  marine metagenome\n",
       "DRR086642   S26     0.024569  marine metagenome"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# pull out just the ones with good scientific names:\n",
    "magsearch3_df = magsearch2_df[~magsearch2_df['ScientificName'].isnull()]\n",
    "perc_non_null = len(magsearch3_df)/len(magsearch2_df)*100\n",
    "print(f\"Of {len(magsearch2_df)} MAGsearch results, {len(magsearch3_df)} have non-null metadata ({perc_non_null:.2f}%)\")\n",
    "magsearch3_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a5c85aa8-b57f-438d-bfda-0e09e2d1011b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4 independent queries in results\n"
     ]
    }
   ],
   "source": [
    "print(f'{len(set(magsearch3_df[\"query\"]))} independent queries in results')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "191385d2-3c38-47cc-93d2-0b8bd1f0bc78",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "FV_DSM_15829_genome     3083\n",
       "FV_PB189-T1-4_genome     890\n",
       "S26                      841\n",
       "C3_T13_0                 112\n",
       "Name: query, dtype: int64"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# how many matches do we have for each query?\n",
    "magsearch3_df[\"query\"].value_counts()#[:20]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c4d3447-2af6-4d9a-9845-445fe6242ef0",
   "metadata": {},
   "source": [
    "### Split Results by Query\n",
    "\n",
    "We two very different queries! Let's split our results into a dataframe that only contains the marine queries.\n",
    "marine_queries = [\"S26\", \"C3_T13_0\"]\n",
    "\n",
    "vg_queries = [\"FV_DSM_15829_genome\", \"FV_PB189-T1-4_genome\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "4d47aac4-ff63-4477-b61c-9233cd9c20a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "marine_queries = [\"S26\", \"C3_T13_0\"]\n",
    "#vg_queries = [\"FV_DSM_15829_genome\", \"FV_PB189-T1-4_genome\"]\n",
    "\n",
    "marine_df = magsearch3_df[magsearch3_df[\"query\"].isin(marine_queries)]\n",
    "#vg_df = magsearch3_df[magsearch3_df[\"query\"].isin(vg_queries)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "3d5d9587-9f48-4701-bcbe-dda688be47d6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['S26', 'C3_T13_0'], dtype=object)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "marine_df[\"query\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "fd1cdc51-8b0b-4878-a930-8cfb34516237",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query</th>\n",
       "      <th>containment</th>\n",
       "      <th>ScientificName</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Run</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DRR086622</th>\n",
       "      <td>S26</td>\n",
       "      <td>0.020896</td>\n",
       "      <td>marine metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRR086623</th>\n",
       "      <td>S26</td>\n",
       "      <td>0.018140</td>\n",
       "      <td>marine metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRR086624</th>\n",
       "      <td>S26</td>\n",
       "      <td>0.019977</td>\n",
       "      <td>marine metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRR086640</th>\n",
       "      <td>S26</td>\n",
       "      <td>0.017451</td>\n",
       "      <td>marine metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRR086642</th>\n",
       "      <td>S26</td>\n",
       "      <td>0.024569</td>\n",
       "      <td>marine metagenome</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          query  containment     ScientificName\n",
       "Run                                            \n",
       "DRR086622   S26     0.020896  marine metagenome\n",
       "DRR086623   S26     0.018140  marine metagenome\n",
       "DRR086624   S26     0.019977  marine metagenome\n",
       "DRR086640   S26     0.017451  marine metagenome\n",
       "DRR086642   S26     0.024569  marine metagenome"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "marine_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad0581ea-4eb8-45d9-94e2-706cfcb6c6a0",
   "metadata": {},
   "source": [
    "# Start looking at the results!\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "f0cfb63c-523e-4cae-8efe-f302b31a25a4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "marine metagenome             167\n",
       "biofilm metagenome            137\n",
       "metagenome                    131\n",
       "seawater metagenome            87\n",
       "sediment metagenome            73\n",
       "aquatic metagenome             64\n",
       "marine sediment metagenome     34\n",
       "oyster metagenome              28\n",
       "synthetic metagenome           19\n",
       "estuary metagenome             19\n",
       "soil metagenome                16\n",
       "wastewater metagenome          15\n",
       "microbial mat metagenome       15\n",
       "viral metagenome               12\n",
       "salt marsh metagenome          11\n",
       "coral reef metagenome           9\n",
       "Escherichia                     9\n",
       "freshwater metagenome           9\n",
       "plastic metagenome              6\n",
       "plant metagenome                6\n",
       "Name: ScientificName, dtype: int64"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# what are the top ScientificNames of the matches?\n",
    "marine_df[\"ScientificName\"].value_counts()[:20]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4cc75c4-c9b2-48e9-a9f2-fa9bea626de8",
   "metadata": {},
   "source": [
    "## Sort Results by Containment\n",
    "\n",
    "The default threshold for containment is 0.01, which means ~1% of the query genome needs to be found in the metagenome for it to be reported. That's not very stringent!\n",
    "\n",
    "First, let's look at the SRA runs that had the **best** containment of our queries:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "95a5e0cd-78e0-457d-ab7e-6f783d6a22cd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query</th>\n",
       "      <th>containment</th>\n",
       "      <th>ScientificName</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Run</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ERR2094172</th>\n",
       "      <td>S26</td>\n",
       "      <td>0.339839</td>\n",
       "      <td>metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6675345</th>\n",
       "      <td>S26</td>\n",
       "      <td>0.245924</td>\n",
       "      <td>gut metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR12112861</th>\n",
       "      <td>S26</td>\n",
       "      <td>0.177268</td>\n",
       "      <td>seawater metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4674711</th>\n",
       "      <td>S26</td>\n",
       "      <td>0.166475</td>\n",
       "      <td>marine metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR2094169</th>\n",
       "      <td>S26</td>\n",
       "      <td>0.143054</td>\n",
       "      <td>metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR12112859</th>\n",
       "      <td>S26</td>\n",
       "      <td>0.134558</td>\n",
       "      <td>seawater metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR2094168</th>\n",
       "      <td>S26</td>\n",
       "      <td>0.133639</td>\n",
       "      <td>metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR2094175</th>\n",
       "      <td>S26</td>\n",
       "      <td>0.115040</td>\n",
       "      <td>metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR3405417</th>\n",
       "      <td>S26</td>\n",
       "      <td>0.107003</td>\n",
       "      <td>synthetic metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR3405420</th>\n",
       "      <td>S26</td>\n",
       "      <td>0.096670</td>\n",
       "      <td>synthetic metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR2185279</th>\n",
       "      <td>S26</td>\n",
       "      <td>0.094374</td>\n",
       "      <td>synthetic metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR1992808</th>\n",
       "      <td>S26</td>\n",
       "      <td>0.094374</td>\n",
       "      <td>synthetic metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR9037724</th>\n",
       "      <td>S26</td>\n",
       "      <td>0.093226</td>\n",
       "      <td>aquatic metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR16548371</th>\n",
       "      <td>S26</td>\n",
       "      <td>0.087486</td>\n",
       "      <td>aquatic metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR16548364</th>\n",
       "      <td>S26</td>\n",
       "      <td>0.081286</td>\n",
       "      <td>aquatic metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR16548366</th>\n",
       "      <td>S26</td>\n",
       "      <td>0.081056</td>\n",
       "      <td>aquatic metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR8497107</th>\n",
       "      <td>S26</td>\n",
       "      <td>0.081056</td>\n",
       "      <td>biofilm metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR3405423</th>\n",
       "      <td>S26</td>\n",
       "      <td>0.079219</td>\n",
       "      <td>synthetic metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR3405421</th>\n",
       "      <td>S26</td>\n",
       "      <td>0.079219</td>\n",
       "      <td>synthetic metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR16548368</th>\n",
       "      <td>S26</td>\n",
       "      <td>0.076005</td>\n",
       "      <td>aquatic metagenome</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            query  containment        ScientificName\n",
       "Run                                                 \n",
       "ERR2094172    S26     0.339839            metagenome\n",
       "SRR6675345    S26     0.245924        gut metagenome\n",
       "SRR12112861   S26     0.177268   seawater metagenome\n",
       "ERR4674711    S26     0.166475     marine metagenome\n",
       "ERR2094169    S26     0.143054            metagenome\n",
       "SRR12112859   S26     0.134558   seawater metagenome\n",
       "ERR2094168    S26     0.133639            metagenome\n",
       "ERR2094175    S26     0.115040            metagenome\n",
       "SRR3405417    S26     0.107003  synthetic metagenome\n",
       "SRR3405420    S26     0.096670  synthetic metagenome\n",
       "ERR2185279    S26     0.094374  synthetic metagenome\n",
       "ERR1992808    S26     0.094374  synthetic metagenome\n",
       "SRR9037724    S26     0.093226    aquatic metagenome\n",
       "SRR16548371   S26     0.087486    aquatic metagenome\n",
       "SRR16548364   S26     0.081286    aquatic metagenome\n",
       "SRR16548366   S26     0.081056    aquatic metagenome\n",
       "SRR8497107    S26     0.081056    biofilm metagenome\n",
       "SRR3405423    S26     0.079219  synthetic metagenome\n",
       "SRR3405421    S26     0.079219  synthetic metagenome\n",
       "SRR16548368   S26     0.076005    aquatic metagenome"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "marine_df.sort_values(by=['containment'], ascending=False)[:20]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29b3619d-46e6-4c06-98d2-3ba1a71f24c6",
   "metadata": {},
   "source": [
    "## Filter Results by Containment\n",
    "\n",
    "We've found (rule of thumb) that 0.2 is a decent value - 20% - indicating some level of stringency. Let's take a look -"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "7749929d-9fde-4051-ba4f-c20a4cc5611e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "query type: marine\n",
      "total matches: 2\n",
      "query: 1\n",
      "metagenomes: 2\n",
      "\n",
      "\n",
      "metagenome        1\n",
      "gut metagenome    1\n",
      "Name: ScientificName, dtype: int64 \n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# let's do some filtering -\n",
    "marine_df2 = marine_df[marine_df['containment'] > 0.2]\n",
    "\n",
    "for name, df in {\"marine\": marine_df2}.items():\n",
    "    print('query type:', name)\n",
    "    print('total matches:', len(df))\n",
    "    print('query:', len(set(df[\"query\"])))\n",
    "    print('metagenomes:', len(set(df.index)))\n",
    "    print(\"\\n\")\n",
    "    print(df[\"ScientificName\"].value_counts()[:20], \"\\n\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "ae09e46a-f64a-420e-a739-5e088665b644",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query</th>\n",
       "      <th>containment</th>\n",
       "      <th>ScientificName</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Run</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ERR2094172</th>\n",
       "      <td>S26</td>\n",
       "      <td>0.339839</td>\n",
       "      <td>metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6675345</th>\n",
       "      <td>S26</td>\n",
       "      <td>0.245924</td>\n",
       "      <td>gut metagenome</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           query  containment  ScientificName\n",
       "Run                                          \n",
       "ERR2094172   S26     0.339839      metagenome\n",
       "SRR6675345   S26     0.245924  gut metagenome"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "marine_df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "43a322a2-aef6-4ea4-8d50-84c58b223b16",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "query type: marine\n",
      "total matches: 58\n",
      "query: 1\n",
      "metagenomes: 58\n",
      "\n",
      "\n",
      "aquatic metagenome       18\n",
      "seawater metagenome      13\n",
      "synthetic metagenome      9\n",
      "metagenome                8\n",
      "biofilm metagenome        5\n",
      "seagrass metagenome       1\n",
      "marine metagenome         1\n",
      "sponge metagenome         1\n",
      "freshwater metagenome     1\n",
      "gut metagenome            1\n",
      "Name: ScientificName, dtype: int64 \n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# only two!! filter less stringently..\n",
    "# let's do some filtering -\n",
    "marine_df3 = marine_df[marine_df['containment'] > 0.05]\n",
    "\n",
    "for name, df in {\"marine\": marine_df3}.items():\n",
    "    print('query type:', name)\n",
    "    print('total matches:', len(df))\n",
    "    print('query:', len(set(df[\"query\"])))\n",
    "    print('metagenomes:', len(set(df.index)))\n",
    "    print(\"\\n\")\n",
    "    print(df[\"ScientificName\"].value_counts()[:20], \"\\n\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "afc2ad4c-42ef-4deb-99c8-f223bd2abdf1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save the 5% containment results to a file\n",
    "marine_df3.to_csv(\"marine_metagenome_results.0.05.containment.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cac862ab-3bcc-4cd0-95b8-02967a551307",
   "metadata": {},
   "outputs": [],
   "source": [
    "# look at just one genome..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "869ea1b9-faee-40d7-a532-ea3ef521a961",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query</th>\n",
       "      <th>containment</th>\n",
       "      <th>ScientificName</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Run</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ERR1992794</th>\n",
       "      <td>C3_T13_0</td>\n",
       "      <td>0.011612</td>\n",
       "      <td>synthetic metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR1992808</th>\n",
       "      <td>C3_T13_0</td>\n",
       "      <td>0.013661</td>\n",
       "      <td>synthetic metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR2092773</th>\n",
       "      <td>C3_T13_0</td>\n",
       "      <td>0.010018</td>\n",
       "      <td>metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR2092774</th>\n",
       "      <td>C3_T13_0</td>\n",
       "      <td>0.013661</td>\n",
       "      <td>metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR2092775</th>\n",
       "      <td>C3_T13_0</td>\n",
       "      <td>0.013206</td>\n",
       "      <td>metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR8497110</th>\n",
       "      <td>C3_T13_0</td>\n",
       "      <td>0.018898</td>\n",
       "      <td>biofilm metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR8497111</th>\n",
       "      <td>C3_T13_0</td>\n",
       "      <td>0.017760</td>\n",
       "      <td>biofilm metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR8497112</th>\n",
       "      <td>C3_T13_0</td>\n",
       "      <td>0.023679</td>\n",
       "      <td>biofilm metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR9037724</th>\n",
       "      <td>C3_T13_0</td>\n",
       "      <td>0.020947</td>\n",
       "      <td>aquatic metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR9045884</th>\n",
       "      <td>C3_T13_0</td>\n",
       "      <td>0.017760</td>\n",
       "      <td>microbial mat metagenome</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>112 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               query  containment            ScientificName\n",
       "Run                                                        \n",
       "ERR1992794  C3_T13_0     0.011612      synthetic metagenome\n",
       "ERR1992808  C3_T13_0     0.013661      synthetic metagenome\n",
       "ERR2092773  C3_T13_0     0.010018                metagenome\n",
       "ERR2092774  C3_T13_0     0.013661                metagenome\n",
       "ERR2092775  C3_T13_0     0.013206                metagenome\n",
       "...              ...          ...                       ...\n",
       "SRR8497110  C3_T13_0     0.018898        biofilm metagenome\n",
       "SRR8497111  C3_T13_0     0.017760        biofilm metagenome\n",
       "SRR8497112  C3_T13_0     0.023679        biofilm metagenome\n",
       "SRR9037724  C3_T13_0     0.020947        aquatic metagenome\n",
       "SRR9045884  C3_T13_0     0.017760  microbial mat metagenome\n",
       "\n",
       "[112 rows x 3 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "c3 = marine_df[marine_df['query'] == \"C3_T13_0\"]\n",
    "c3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "7d7101f2-d0cc-4e3b-a5c2-4e966bc79ddf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query</th>\n",
       "      <th>containment</th>\n",
       "      <th>ScientificName</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Run</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR8497107</th>\n",
       "      <td>C3_T13_0</td>\n",
       "      <td>0.030282</td>\n",
       "      <td>biofilm metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR16548364</th>\n",
       "      <td>C3_T13_0</td>\n",
       "      <td>0.030055</td>\n",
       "      <td>aquatic metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR16548371</th>\n",
       "      <td>C3_T13_0</td>\n",
       "      <td>0.030055</td>\n",
       "      <td>aquatic metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR16548366</th>\n",
       "      <td>C3_T13_0</td>\n",
       "      <td>0.028689</td>\n",
       "      <td>aquatic metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4592245</th>\n",
       "      <td>C3_T13_0</td>\n",
       "      <td>0.028005</td>\n",
       "      <td>seagrass metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR16548368</th>\n",
       "      <td>C3_T13_0</td>\n",
       "      <td>0.027778</td>\n",
       "      <td>aquatic metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR16548367</th>\n",
       "      <td>C3_T13_0</td>\n",
       "      <td>0.024135</td>\n",
       "      <td>aquatic metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR8497102</th>\n",
       "      <td>C3_T13_0</td>\n",
       "      <td>0.023907</td>\n",
       "      <td>biofilm metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR8497112</th>\n",
       "      <td>C3_T13_0</td>\n",
       "      <td>0.023679</td>\n",
       "      <td>biofilm metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR16548375</th>\n",
       "      <td>C3_T13_0</td>\n",
       "      <td>0.023224</td>\n",
       "      <td>aquatic metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR16548373</th>\n",
       "      <td>C3_T13_0</td>\n",
       "      <td>0.022996</td>\n",
       "      <td>aquatic metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR3428544</th>\n",
       "      <td>C3_T13_0</td>\n",
       "      <td>0.022541</td>\n",
       "      <td>metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR8497108</th>\n",
       "      <td>C3_T13_0</td>\n",
       "      <td>0.022313</td>\n",
       "      <td>biofilm metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR16548390</th>\n",
       "      <td>C3_T13_0</td>\n",
       "      <td>0.021858</td>\n",
       "      <td>aquatic metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR8497105</th>\n",
       "      <td>C3_T13_0</td>\n",
       "      <td>0.021858</td>\n",
       "      <td>biofilm metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR12681089</th>\n",
       "      <td>C3_T13_0</td>\n",
       "      <td>0.021403</td>\n",
       "      <td>plant metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR16548360</th>\n",
       "      <td>C3_T13_0</td>\n",
       "      <td>0.021175</td>\n",
       "      <td>aquatic metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR16548358</th>\n",
       "      <td>C3_T13_0</td>\n",
       "      <td>0.021175</td>\n",
       "      <td>aquatic metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR9037724</th>\n",
       "      <td>C3_T13_0</td>\n",
       "      <td>0.020947</td>\n",
       "      <td>aquatic metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6675345</th>\n",
       "      <td>C3_T13_0</td>\n",
       "      <td>0.020947</td>\n",
       "      <td>gut metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR16548374</th>\n",
       "      <td>C3_T13_0</td>\n",
       "      <td>0.020719</td>\n",
       "      <td>aquatic metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6854605</th>\n",
       "      <td>C3_T13_0</td>\n",
       "      <td>0.020036</td>\n",
       "      <td>biofilm metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR8497106</th>\n",
       "      <td>C3_T13_0</td>\n",
       "      <td>0.019809</td>\n",
       "      <td>biofilm metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4592243</th>\n",
       "      <td>C3_T13_0</td>\n",
       "      <td>0.019809</td>\n",
       "      <td>seagrass metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR16548365</th>\n",
       "      <td>C3_T13_0</td>\n",
       "      <td>0.019809</td>\n",
       "      <td>aquatic metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4592244</th>\n",
       "      <td>C3_T13_0</td>\n",
       "      <td>0.019353</td>\n",
       "      <td>macroalgae metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR8497110</th>\n",
       "      <td>C3_T13_0</td>\n",
       "      <td>0.018898</td>\n",
       "      <td>biofilm metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR16548363</th>\n",
       "      <td>C3_T13_0</td>\n",
       "      <td>0.018898</td>\n",
       "      <td>aquatic metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR16548370</th>\n",
       "      <td>C3_T13_0</td>\n",
       "      <td>0.018670</td>\n",
       "      <td>aquatic metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR6877521</th>\n",
       "      <td>C3_T13_0</td>\n",
       "      <td>0.018670</td>\n",
       "      <td>biofilm metagenome</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                query  containment         ScientificName\n",
       "Run                                                      \n",
       "SRR8497107   C3_T13_0     0.030282     biofilm metagenome\n",
       "SRR16548364  C3_T13_0     0.030055     aquatic metagenome\n",
       "SRR16548371  C3_T13_0     0.030055     aquatic metagenome\n",
       "SRR16548366  C3_T13_0     0.028689     aquatic metagenome\n",
       "ERR4592245   C3_T13_0     0.028005    seagrass metagenome\n",
       "SRR16548368  C3_T13_0     0.027778     aquatic metagenome\n",
       "SRR16548367  C3_T13_0     0.024135     aquatic metagenome\n",
       "SRR8497102   C3_T13_0     0.023907     biofilm metagenome\n",
       "SRR8497112   C3_T13_0     0.023679     biofilm metagenome\n",
       "SRR16548375  C3_T13_0     0.023224     aquatic metagenome\n",
       "SRR16548373  C3_T13_0     0.022996     aquatic metagenome\n",
       "ERR3428544   C3_T13_0     0.022541             metagenome\n",
       "SRR8497108   C3_T13_0     0.022313     biofilm metagenome\n",
       "SRR16548390  C3_T13_0     0.021858     aquatic metagenome\n",
       "SRR8497105   C3_T13_0     0.021858     biofilm metagenome\n",
       "SRR12681089  C3_T13_0     0.021403       plant metagenome\n",
       "SRR16548360  C3_T13_0     0.021175     aquatic metagenome\n",
       "SRR16548358  C3_T13_0     0.021175     aquatic metagenome\n",
       "SRR9037724   C3_T13_0     0.020947     aquatic metagenome\n",
       "SRR6675345   C3_T13_0     0.020947         gut metagenome\n",
       "SRR16548374  C3_T13_0     0.020719     aquatic metagenome\n",
       "SRR6854605   C3_T13_0     0.020036     biofilm metagenome\n",
       "SRR8497106   C3_T13_0     0.019809     biofilm metagenome\n",
       "ERR4592243   C3_T13_0     0.019809    seagrass metagenome\n",
       "SRR16548365  C3_T13_0     0.019809     aquatic metagenome\n",
       "ERR4592244   C3_T13_0     0.019353  macroalgae metagenome\n",
       "SRR8497110   C3_T13_0     0.018898     biofilm metagenome\n",
       "SRR16548363  C3_T13_0     0.018898     aquatic metagenome\n",
       "SRR16548370  C3_T13_0     0.018670     aquatic metagenome\n",
       "SRR6877521   C3_T13_0     0.018670     biofilm metagenome"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "c3.sort_values(by=['containment'], ascending=False)[:30]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9e11fe3-a8c7-4aea-967d-5d4a3de79433",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
